ddl/placement_policy_test.go

```diff
@@ -427,3 +427,114 @@ func (s *testDBSuite6) TestDropPlacementPolicyInUse(c *C) {
 		c.Assert(err.Error(), Equals, fmt.Sprintf("[ddl:8241]Placement policy '%s' is still in use", policyName))
 	}
 }
+
+func testGetPolicyByName(c *C, ctx sessionctx.Context, name string, mustExist bool) *model.PolicyInfo {
+	dom := domain.GetDomain(ctx)
+	// Make sure the table schema is the new schema.
+	err := dom.Reload()
+	c.Assert(err, IsNil)
+	po, ok := dom.InfoSchema().PolicyByName(model.NewCIStr(name))
+	if mustExist {
+		c.Assert(ok, Equals, true)
+	}
+	return po
+}
+
+func testGetPolicyDependency(storage kv.Storage, name string) []int64 {
+	ids := make([]int64, 0, 32)
+	err1 := kv.RunInNewTxn(context.Background(), storage, false, func(ctx context.Context, txn kv.Transaction) error {
+		t := meta.NewMeta(txn)
+		dbs, err := t.ListDatabases()
+		if err != nil {
+			return err
+		}
+		for _, db := range dbs {
+			tbls, err := t.ListTables(db.ID)
+			if err != nil {
+				return err
+			}
+			for _, tbl := range tbls {
+				if tbl.PlacementPolicyRef != nil && tbl.PlacementPolicyRef.Name.L == name {
+					ids = append(ids, tbl.ID)
+				}
+			}
+		}
+		return nil
+	})
+	if err1 != nil {
+		return []int64{}
+	}
+	return ids
+}
+
+func (s *testDBSuite6) TestPolicyCacheAndPolicyDependencyCache(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop placement policy if exists x")
+
+	// Test policy cache.
+	tk.MustExec("create placement policy x primary_region=\"r1\" regions=\"r1,r2\" schedule=\"EVEN\";")
+	po := testGetPolicyByName(c, tk.Se, "x", true)
+	c.Assert(po, NotNil)
+	tk.MustQuery("show placement").Check(testkit.Rows("POLICY x PRIMARY_REGION=\"r1\" REGIONS=\"r1,r2\" SCHEDULE=\"EVEN\" SCHEDULED"))
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int) placement policy \"x\"")
+	tbl := testGetTableByName(c, tk.Se, "test", "t")
+
+	// Test policy dependency cache.
+	dependencies := testGetPolicyDependency(s.store, "x")
+	c.Assert(dependencies, NotNil)
+	c.Assert(len(dependencies), Equals, 1)
+	c.Assert(dependencies[0], Equals, tbl.Meta().ID)
+
+	tk.MustExec("drop table if exists t2")
+	tk.MustExec("create table t2 (a int) placement policy \"x\"")
+	tbl2 := testGetTableByName(c, tk.Se, "test", "t2")
+
+	dependencies = testGetPolicyDependency(s.store, "x")
+	c.Assert(dependencies, NotNil)
+	c.Assert(len(dependencies), Equals, 2)
+	in := func() bool {
+		for _, one := range dependencies {
+			if one == tbl2.Meta().ID {
+				return true
+			}
+		}
+		return false
+	}
+	c.Assert(in(), Equals, true)
+
+	// Test drop policy can't succeed cause there are still some table depend on them.
+	_, err := tk.Exec("drop placement policy x")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[ddl:8241]Placement policy 'x' is still in use")
+
+	// Drop depended table t firstly.
+	tk.MustExec("drop table if exists t")
+	dependencies = testGetPolicyDependency(s.store, "x")
```

Comment:
I'm not very understand what is the case testing for. I think dependency is a computed value, so if the table will not exists after dropping, it can be guaranteed relationship is right.

Maybe `getPlacementPolicyDependedObjectsIDs` is what we really want to test?

Reply:
I just want to make it explicit by fetching it up and and say: "look! there 2 tables depend on that policy"

Otherwise, it's kind of black box handled by ddl interface. the logic is same like `getPlacementPolicyDependedObjectsIDs`